src/builtins.c

```diff
@@ -342,24 +363,34 @@ static void stream_time( const struct server *serp )
 
 static void dgram_time( const struct server *serp )
 {
-   char     buf[ 1 ] ;
    unsigned char time_buf[4];
    union xsockaddr lsin ;
-   socklen_t       sin_len = 0 ;
    int             fd      = SERVER_FD( serp ) ;
    ssize_t         val;
+   char            cmbuf[ CMSG_SPACE( sizeof( struct in6_pktinfo ) ) ];
+   struct msghdr   mh = {
+      .msg_name = SA( &lsin ),
+      .msg_control = cmbuf,
+      .msg_controllen = sizeof( cmbuf ),
+   };
 
    if ( SC_IPV4( SVC_CONF( SERVER_SERVICE( serp ) ) ) ) 
-      sin_len = sizeof( struct sockaddr_in );
+      mh.msg_namelen = sizeof( struct sockaddr_in );
    else if ( SC_IPV6( SVC_CONF( SERVER_SERVICE( serp ) ) ) ) 
-      sin_len = sizeof( struct sockaddr_in6 );
+      mh.msg_namelen = sizeof( struct sockaddr_in6 );
 
-   val = recvfrom( fd, buf, sizeof( buf ), 0, (struct sockaddr *)( &lsin ), &sin_len );
+   val = recvmsg( fd, &mh, 0 );
    if ( val == (ssize_t)-1 )
       return ;
 
    time_protocol( time_buf ) ;
-   (void) sendto( fd, (char *) time_buf, 4, 0, SA( &lsin ), sin_len ) ;
+
+   struct iovec iov[] = {
+      { .iov_base = time_buf, .iov_len = sizeof(time_buf) },
+   };
+   mh.msg_iov = iov;
+   mh.msg_iovlen = sizeof( iov ) / sizeof( struct iovec );
+   (void) sendmsg( fd, &msg, 0 );
```

Comment:
Oops, just noticed a nasty typo here. Should be `&mh`. I have rewritten this patch a few too many times...